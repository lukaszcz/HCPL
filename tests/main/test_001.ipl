
let power = \a \b {
   if (b /= 0) then {
      let pom = \ta \tb \b {
         if b > 1 then
            pom (ta * ta)
                (if b mod 2 /= 0 then tb * ta else tb)
                (b div 2)
         else
            ta * tb
      };
      pom a 1 b
   } else {
      1
   }
};

print (power 3 4);
print (power 2 16);
print (power 5 7);

let powerl = \a \b {
   if (b /= 0) then {
      let pom = \&ta \&tb \&b {
         if b > 1 then
            pom (ta * ta)
                (if b mod 2 /= 0 then tb * ta else tb)
                (b div 2)
         else
            ta * tb
      };
      pom a 1 b
   } else {
      1
   }
};

print (powerl 3 4);
print (powerl 2 16);
print (powerl 5 7);

let power2 = \a \b {
   if b = 0 then 1
   else if b mod 2 = 1 then a * (power2 (a * a) (b div 2))
   else power2 (a * a) (b div 2);
};

/*
 * This is a big comment.
 *  Tralala!!!
 */

print (power2 3 4);
print (power2 2 16);
print (power2 5 7);

let power2l = \&a \&b {
   if b = 0 then 1
   else if b mod 2 = 1 then a * (power2l (a * a) (b div 2))
   else power2l (a * a) (b div 2);
};

print (power2l 3 4);
print (power2l 2 16);
print (power2l 5 7);
