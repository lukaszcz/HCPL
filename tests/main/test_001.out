81
65536
78125
81
65536
78125
81
65536
78125
81
65536
78125
